onnxruntime/test/shared_lib/utils.h

```diff
@@ -39,9 +39,15 @@ struct MyCustomOp : Ort::CustomOpBase<MyCustomOp, MyCustomKernel> {
     // Both the inputs need to be necessarily of float type
     return ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT;
   };
+  OrtCustomOpInputOutputCharacteristic GetInputCharacteristic(size_t /*index*/) const {
+    return OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_NONOPTIONAL;
```

Comment:
Can the default implementation be defined in a base class? Otherwise any existing custom op will break with an update of ORT that has these new methods won't they?

Reply:
Yes, it will break if trying to re-compile old custom op with latest ORT. I thought about introducing a default implementation, but we will have to do something similar each time we make a similar change (it is setting a precedent). I thought it was okay to introduce source incompatibility ?

Already compiled apps (using older APIs than version 7) will not break and preserve ABI as we won't invoke the callback.

Reply:
Is it a compiliation error if you update ORT or will it be a seg fault at runtime if the function isn't implemented? 

Does [https://github.com/microsoft/ort-customops](https://github.com/microsoft/ort-customops) need updating once this is checked in @wenbingl?

---
In reply to: [578213062](https://github.com/microsoft/onnxruntime/pull/6727#discussion_r578213062) [](ancestors = 578213062)

Reply:
It will just be a compilation error. Since this won't be included in this release, I plan to update the minimum custom op version for this feature to be 8 (custom op version is the same as the C API version). For 7 and below, it won't segfault as we simply won't invoke the method. For 8 and above (i.e.) anyone using the latest headers to compile their project will result in compilation error unless they implement the method(s). That being said, I can add a default implementation so that older projects will compile with latest headers without any problem.

Reply:
Added a default implementation for the new methods - just like the one there already was to fetch the execution provider for the custom op